Bug 1181571 - Fix lockscreen delay issue

cr opened the pull request:
This is a manually verified fix and prototype tests.

I need some help with the tests. I have trouble testing code paths that resolve asynchronously after method calls have returned, for example in setTimeout() or promises. Is there a test pattern for this case?

Commit:
Fixing lockscreen delay bypass by rewriting core logic

This is fixing bug 1181571 where lockscreen delay on repeated pass code
failure could be trivially bypassed by pressing the home button. The
faulty behavior was caused by a limitation of the previous architecture
which was only making the delay occurr *after* a wrong pass code had
been entered. Consequently, whenever the input pad would be restarted,
like after closing it with the home button or when a screen timeout
occurred, the pending delay would not persist and pass code entry would
be accepted right away.

The fix made it necessary to rewrite parts of lockscreen's messaging
and state handling. Previous changes in architecture caused pass code
failure counting and delay timer handling to be maintained in two
places: in LockScreen (for pass code checking) and in LockScreenInputPad
(for displaying error state to the user). Logic error even caused the
two states to get out of sync under certain conditions.

Now state and timeout related to pass code checking are maintained
exclusively by LockScreen, which signals LockScreenInputManager with
three distinct messages of failure (on wrong pass code), reset (after
error delay has passed), and success (on correct pass code).

LockScreenInputPad now is always aware of a passcode error delay ongoing
in LockScreen, every UI update is rendered according to this state, and
key input is ignored when necessary.

Commit:
Add test prototypes for pass code handling

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fbe8085116264cb3dd9d28e159eac98b687d74dd

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -117,11 +115,11 @@
         evt.preventDefault();
         this.handlePassCodeInput(key);
-      break;
+        // .handlePassCodeInput() triggers .updatePassCode()
```
Remove the commented code.

cr replied:
It's not code, just a comment, and I think it's necessary to follow the logic without looking any further.

Commit:
Fix comment

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=35fb616d74fa4bfea1846d6e3274ba2fb6389116

snowmantw commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -92,20 +91,19 @@
+        this.states.passCodeErrorTimeoutPending = true;
+        this.updatePassCodeUI();
+        break;
+      case 'lockscreen-notify-passcode-validationreset':
```
Although I can understand the possible reason why we need to have two events for the same effects, I think it is better to leave some comments about that. Since according the event flow it's quite strange without knowing the current status of LockScreen:

  validationfailed ----> validationreset  === validationsuccess

So I think we should add few lines here, to explain when the reset would be fired. That is, after the failure and we need to reset the UI.

Commit:
Add comment on event flow

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f51d134acad733df034e2ff268bf8f7b61fc3c31

snowmantw commented on apps/system/test/unit/lockscreen/lockscreen_test.js:
```diff
@@ -297,6 +297,75 @@ suite('system/LockScreen >', function() {
+    //TODO: test validationreset event triggered async after timeout
+
+    /* TODO: promise resolves after checkPassCode() returns.
+     * TODO: how can we check this?
```
You can use `test('...', function(done) {...})` and then put the `done` in the promise's next step. Namely:

    test('...', function(done) {
        function() {
            return Promise.resolve(true).then(done).catch(done);
            // The first `done` ends this  asynchronous test.
            // The second `done` catch the error and make the test failed to present that
        }
    } );

cr replied:
This results in Error: done() invoked with non-Error: true.

cr replied:
When I change from ```.then(done)``` to ```.then(function(){done()})```,  the result is *Error: done() called multiple times* although I can prove by adding a ```console.log()```-call that done is called exactly once, but after the test reported having failed. Looks like there's an implicit ```done()``` somewhere after the test returns.

I suspect that the suggested pattern may not be the solution for this particular case.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ac16d3cf4b0d510b257571eb5405c694260b6f02

cr commented:
Rebased to today's master.

Commit:
Finalizing tests

Tests now working. Solved by creating a setTimeout stub that forces
synchronicity. Also removed onPasscodeValidationFailed trigger check
on wrong pass code, because I can't make the promise resolve before
the test ends without code changes in the test subject.

Commit:
Overhaul tests for new internal InputPad logic

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bed700e1d592917e18b41f43746df8780a4e8d99

Commit:
Fix TypeError: this.emergencyCallBtn is null

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=09263c146580291e539ca7ccbe420e95face6080

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=785b58fd5d5b08208f52a648f5d59ba9204eddc1

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=28b85c0b4bcd8f3b8144121393ab24685d9272c1
